imap_processing/codice/codice_l2.py

```diff
@@ -866,6 +907,133 @@ def process_hi_sectored(dependencies: ProcessingInputCollection) -> xr.Dataset:
     return l2_dataset
 
 
+def process_hi_direct_events(dependencies: ProcessingInputCollection) -> xr.Dataset:
+    """
+    Process the hi-direct-events L1A dataset to convert variables to physical units.
+
+    See section 11.2.1 of the CoDICE algorithm document for details.
+
+    Parameters
+    ----------
+    dependencies : ProcessingInputCollection
+        The collection of processing input files.
+
+    Returns
+    -------
+    xarray.Dataset
+        The updated L2 dataset with variables converted to physical units.
+    """
+    file_path = dependencies.get_file_paths(descriptor="hi-direct-events")[0]
+    l1a_dataset = load_cdf(file_path)
+
+    # Update global CDF attributes
+    cdf_attrs = ImapCdfAttributes()
+    cdf_attrs.add_instrument_global_attrs("codice")
+    cdf_attrs.add_instrument_variable_attrs("codice", "l2-hi-direct-events")
+
+    l2_dataset = l1a_dataset
+    # Load energy table and tof table needed for conversions
+    energy_table, tof_table = get_hi_de_luts(dependencies)
+
+    # Convert from position to elevation angle in degrees relative to the spacecraft
+    # axis
+    elevation_angle_shape = l2_dataset["ssd_id"].shape
+    ssd_id_flat = l2_dataset["ssd_id"].values.ravel()
+    elevation_angle = np.array(
+        [SSD_ID_TO_ELEVATION.get(id, np.nan) for id in ssd_id_flat]
+    ).reshape(elevation_angle_shape)
+
+    l2_dataset["elevation_angle"] = (
+        l2_dataset["ssd_id"].dims,
+        elevation_angle.astype(np.float32),
+    )
+    # Calculate ssd energy in meV
+    gain_flat = l2_dataset["gain"].values.ravel()
+    ssd_energy_flat = l2_dataset["ssd_energy"].values.ravel()
+    # Initialize ssd_energy with nans
+    ssd_energy = np.full(ssd_energy_flat.shape, np.nan)
+    valid_mask = (
+        (np.isin(gain_flat, list(GAIN_ID_TO_STR.keys())))
+        & (ssd_id_flat <= 15)
+        & (ssd_energy_flat != len(energy_table))
+    )
+    # The columns are organized in order of id and gains
+    # E.g. ssd 0 - LG, ssd 0 - MG, ssd 0 - HG, ssd 1 - LG, ssd 1 - MG, ssd 1 - HG, ...
+    cols = ssd_id_flat * 3 + (gain_flat - 1)
+    ssd_energy[valid_mask] = energy_table[ssd_energy_flat[valid_mask], cols[valid_mask]]
+    l2_dataset["ssd_energy"].data = ssd_energy.reshape(
+        l2_dataset["ssd_energy"].shape
+    ).astype(np.float32)
```

Comment:
This flattening then reshaping seems like we could use xarray dimension broadcasting better here somehow to have things work for us without needing to jump through these hoops. I'm not sure exactly what that would look like or how hard it would be to make that work, but with your use of `data.where()` in other PRs as well it seems doable to me. Maybe chalk this one up to a future refactor in Phase E though unless something jumps out to you immediately.